petastorm/spark/spark_dataset_converter.py

```diff
@@ -0,0 +1,289 @@
+        cached_df_meta = CachedDataFrameMeta.create_cached_dataframe(
+            df, parent_cache_dir_url, parquet_row_group_size_bytes,
+            compression_codec)
+        _cache_df_meta_list.append(cached_df_meta)
+        return cached_df_meta.cache_dir_url
+
+
+def _materialize_df(df, parent_cache_dir, parquet_row_group_size_bytes,
+                    compression_codec):
+    uuid_str = str(uuid.uuid4())
+    save_to_dir = os.path.join(parent_cache_dir, uuid_str)
+
+    df.write \
+        .option("compression", compression_codec) \
+        .option("parquet.block.size", parquet_row_group_size_bytes) \
+        .parquet(save_to_dir)
+    atexit.register(_delete_cache_data_atexit, save_to_dir)
+
+    return save_to_dir
+
+
+def make_spark_converter(
+        df,
+        cache_dir_url=None,
+        parquet_row_group_size_bytes=DEFAULT_ROW_GROUP_SIZE_BYTES,
+        compression=None):
+    """
+    Convert a spark dataframe into a :class:`SparkDatasetConverter` object.
+    It will materialize a spark dataframe to a `cache_dir_url`.
+    The returned `SparkDatasetConverter` object will hold the materialized
+    dataframe, and can be used to make one or more tensorflow datasets or
+    torch dataloaders.
+
+    :param df: The :class:`DataFrame` object to be converted.
+    :param cache_dir_url: A URL string denoting the parent directory to store
+        intermediate files. Default None, it will fallback to the spark config
+        "petastorm.spark.converter.defaultCacheDirUrl".
+    :param parquet_row_group_size_bytes: An int denoting the number of bytes
+        in a parquet row group.
+    :param compression: True or False, specify whether to apply compression.
+        Default None. If None, it will automatically choose the best way.
+
+    :return: a :class:`SparkDatasetConverter` object that holds the
+        materialized dataframe and can be used to make one or more tensorflow
+        datasets or torch dataloaders.
+    """
+    if cache_dir_url is None:
+        cache_dir_url = _get_spark_session().conf \
+            .get("petastorm.spark.converter.defaultCacheDirUrl", None)
+
+    if cache_dir_url is None:
+        raise ValueError(
+            "Please specify the parameter cache_dir_url denoting the parent "
+            "directory to store intermediate files, or set the spark config "
+            "`petastorm.spark.converter.defaultCacheDirUrl`.")
+
+    cache_dir_url = _normalize_dir_url(cache_dir_url)
+
+    if compression is None:
+        # TODO: Improve default behavior to be automatically choosing the
+        #  best way.
+        compression_codec = "uncompressed"
+    elif compression:
+        compression_codec = "snappy"
+    else:
+        compression_codec = "uncompressed"
+
+    dataset_cache_dir_url = _cache_df_or_retrieve_cache_data_url(
+        df, cache_dir_url, parquet_row_group_size_bytes, compression_codec)
+    dataset_size = _get_spark_session().read.parquet(dataset_cache_dir_url).count()
```

Comment:
Just to be sure: in Spark, is this a metadata check?  I am curious about the overhead of creating a DataFrame even if it doesn't process the whole dataset.  It may be faster to just use PyArrow here.

Reply:
Not very sure (depends on spark sql optimizer). Use PyArrow to read count may be a safe way.

Reply:
The overhead of creating a DataFrame is small (just generate an sql plan).